clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -576,16 +590,21 @@ private void addToCompletedReceives() {
                 KafkaChannel channel = entry.getKey();
                 if (!channel.isMute()) {
                     Deque<NetworkReceive> deque = entry.getValue();
-                    NetworkReceive networkReceive = deque.poll();
-                    this.completedReceives.add(networkReceive);
-                    this.sensors.recordBytesReceived(channel.id(), networkReceive.payload().limit());
+                    addToCompletedReceives(channel, deque, false);
```

Comment:
By the way, another way to structure would be to have an `addAllToCompletedReceives` which delegates on each iteration to an `addToCompletedReceives` which adds only a single staged receive. I tend to prefer not having boolean flags as function arguments since they tend to convolute the logic, but feel free to ignore if you don't think it's any better.